### PR TITLE
stm32/mboot: Add verification to fsload / dfu support

### PR DESCRIPTION
**Pull request by andrewleech**
This PR extends the fsload support in stm32/mboot which is used to flash a dfu file stored on an internal filesystem.
* Add CRC check for all dfu files before starting to erase/write.
* Add signature check for entire packed dfu files before starting to erase/write.

#### Background
I'm running into an intermittent issue when using mboot fsload support, where my app saves a dfu file to an LFS2 partition before handing it over to mboot to flash. I'm using the packed (signed/encrypted) dfu files.

In our application code I verify the crc in the dfu file after it's written to flash, before starting mboot with the details for the location of the dfu file. As such I know the file is initially transferred and stored correctly.

Most of the time it works, but under certain circumstances mboot fails part way through, it validated and flashed the first few chunks in the dfu file, then gets to one it rejects due to it not matching the signature.

This particular dfu is not holding firmware, but in fact is targeting a separate partition in an external flash. As such when it fails, the firmware itself is still intact and the device will happily reboots back to application code. Interestingly the partition the DFU file was stored in now reports it cannot mount, so that partition has been corrupted somehow. It all sounds like there's a problem with the address / partitions / overlap / erase or something - but the intermittent nature of this is making it hard to track down.

In the process of diagnosing the root cause of this problem, I noticed the CRC check in fsload was never implemented, so that has been added here in one of the commits.

fsload already had a two-pass approach to handling dfu files, it'd read/walk through it first to check the dfu structure is correct, before resetting the file pointer and read/walking through it again to flash.
The second change included here is to extend this first pass for packed files to also run the signature check on each section, therefore checking the signatures on the entire file match before it starts erasing/flashing anything.

### CODE REVIEW

File: ports/stm32/mboot/fsload.c [hunk 1]
```diff
@@ -177,8 +185,13 @@ static int fsload_program_file(bool write_to_flash) {
     if (res != 16) {
         return -MBOOT_ERRNO_DFU_READ_ERROR;
     }
-
-    // TODO validate CRC32
+    // Don't include final 4 bytes; that's the expected crc.
```
Comment (dpgeorge) [inline ports/stm32/mboot/fsload.c:188]: I think if you do include the final 4 bytes in the calculation, then the output CRC should be expected to be 0 (or 0xffffffff).

Comment (andrewleech) [reply ports/stm32/mboot/fsload.c:188]: I'm pretty sure that only works with the XFER form of crc, not the type/polynomial used in dfu.

Comment (andrewleech) [reply ports/stm32/mboot/fsload.c:188]: Turns out you're probably right! For ease of testing I used a copy of fwupdate.py https://github.com/micropython/micropython/blob/2ea21abae0358a46c0144f014f2e5891afe03dfd/ports/stm32/mboot/fwupdate.py#L78 I already had with the crc function split out:
``` py
def compute_crc(f):
    """Computes the CRC32 value for the file handle passed in,
    minus the final 4 bytes (which hold the checksum itself)."""
    import ubinascii
    f.seek(0, 2)  # os.SEEK_END
    length = f.tell()
    f.seek(0)

    file_crc = 0
    chunk = 16 * 1024
    buff = bytearray(chunk)
    while length >= chunk + 4:
        read = f.readinto(buff)
        length -= read
        file_crc = ubinascii.crc32(buff, file_crc)

    read = f.readinto(buff)
    final = memoryview(buff)[0 : read - 4]
    file_crc = ubinascii.crc32(final, file_crc)
    file_crc = 0xFFFFFFFF & -file_crc - 1

    expected_crc = ubinascii.hexlify(bytes(reversed(memoryview(buff)[read-4:read]))).decode()
    print("calc: 0x%04x, expect: 0x%s" % (file_crc, expected_crc))

    f.seek(0)
    return file_crc
```

```
>>> f=open('PYBD-SF6-20220117-v1.18.dfu', 'rb')
>>> compute_crc(f)
calc: 0x7b37f517, expect: 0x7b37f517
2067264791
>>>
```

Changing that to include the crc in the calc:
``` py
    read = f.readinto(buff)
<   final = memoryview(buff)[0 : read - 4]
>   final = memoryview(buff)[0 : read]
    file_crc = ubinascii.crc32(final, file_crc)
```

Gives
```
>>> compute_crc(f)
calc: 0x0000, expect: 0x7b37f517
0
>>>
```

Comment (dpgeorge) [reply ports/stm32/mboot/fsload.c:188]: Yay!  I did check this also using fwupdate.py, which I wanted to push changes for.

**Comment by dpgeorge** [comment]
Merged in bc856a1e292a287962d629a074260a4cebcb000d and bb0ca00b77b3ccb77bad857373df39556612444a with the following changes:
- the last 4 bytes of the file are included in the CRC calculation, so the CRC32 comes out as 0 on success
- removed the use of redundant `^ 0xffffffff` in the CRC calculation

Tested on PYBD_SF6.

**Pull request closed without merging**